### ISSUE TITLE
IBM XGA Emulation fails with open-source XGAKIT as well

### ISSUE THREAD

**Issue by matt713** [86Box/86Box#2983]
**Describe the bug**
Trying to run XGAKIT gives error:

**`...but you have no XGA adapter`**

**To Reproduce**
1. Setup PS2 Machine in 86Box with XGA option checked
2. install DOS
3. download XGAKIT:
https://ardent-tool.com/video/XGAKIT.ZIP
4. run it 

**Expected behavior**
XGA Kit Demo to run without error message that there is no XGA adapter present.

**Desktop (please complete the following information):**
 - OS: Win 7 x64
 - 86Box version: tested on "3.11 build 4311"
 - Build information: Regular build, AMD64 Release 3.11 version

**Additional context**
Contrary to IBM's XGAAIDOS.SYS, which is not Open-Source and @TC1995 said here:

https://github.com/86Box/86Box/issues/2968#issuecomment-1353478837

it's currently unknown why it fails, XGAKIT is open-source and its:

**`int xga_detect()`**

function source code reveals all exact checks that are done for XGA detection - many of those checks seem to me (as not an expert on IBM PS2 system hardware architecture) are not even XGA-related, but rather MCA-bus-related.  

So, fixing 86Box to working with XGAKIT, i.e. its XGA-check to pass most likely  will greatly improve compatibility with IBM's XGAAIDOS.SYS if not even fully fix that as well, i.e. we may even have two-birds-with-one-stone solution.

In any way, fixing XGAKIT seems like logical first step, because it's Open-Source rather than more complicated situation with IBM's XGAAIDOS.SYS.

**Comment by matt713** [comment]
just to add that XGAKIT source code builds without errors with "Microsoft Macro Assembler 5.10" and Microsoft C 6.0 - that allows its code to be modified, if that's necessary.

**Comment by matt713** [comment]
OK, Good news! I fixed it myself - it wasn't hard to follow the x86 ASM code, especially after my previous comment, i.e. after be able to modify XGAKIT source code and do debug from its side.

So, now XGAKIT starts:
![xgakit_starts_ok](https://user-images.githubusercontent.com/59395613/208888600-89644808-d53b-44f5-9980-54bffdfec300.png)

and 86box XGA Emu even passes 1 out of 8 tests:
![xgakit_test1_pass_ok](https://user-images.githubusercontent.com/59395613/208888877-3b1c44af-e8fb-4172-9659-1d4b296a95d1.png)

 I don't know if other 7 XGAKIT tests are supposed to work, i.e. it's just the current 86box XGA Emu state or there are more bugs.

Later, when have more time, I will provide details and so 86Box developers can fix it from the 86box XGA Emu side, because fixing it from XGAKIT side (as I did) is wrong, i.e. it's just to prove what's wrong in 86box XGA emu at the moment.

**Comment by Cacodemon345** [comment]
Did you test it with XGA2 or XGA1?

**Comment by TC1995** [comment]
Does it expect the 1mb aperture to be enabled?

**Comment by matt713** [comment]
@Cacodemon345
> Did you test it with XGA2 or XGA1?

you can see screenshot of my setup here:

https://github.com/86Box/86Box/issues/2968#issuecomment-1353472691

so, as you can see it's XGA-2 with 1MB Video RAM.

@TC1995
> Does it expect the 1mb aperture to be enabled?
I have 1mb aperture enabled - see my reply to @Cacodemon345 above. So, there are 3 problems with XGAKIT I can Identify:

**PR1 :** what I reported initially in this Issue, i.e. XGAKIT find "no XGA adapter"
**PR2 :** even IBM BIOS/Reference Diskette detects the XGA card as 1MB VRAM, XGAKIT fails and reports 512MB
**PR3:** XGAKIT function xga_mode() fails to set any Graphic mode

So, one by one:

**about PR1 :** what I already understood and 86box code needs to be patched - in very short the problem is "Display ID and Comparator Register (Index 52)" value. So, 86Box returns 0xff, but that's wrong. I was not find able to find any document that describes the meaning of the bits in that register, but experimentally found out that:

 **52H bit 1 :** '1' means Color Monitor, '0' means Monochrome Monitor
 **52H bit 2 :** '1' means Low-Resolution Monitor, '0' means High-Resolution Monitor

That's why I hard-coded value for "Index 52H Register" of 0xfb, i.e. 1111 1 01 1, because as per above:

* bit1 is OK to be '1', i.e. Color, but bit2 needs to be '0' for High-Rez, also values of  bit0-to-bit3 of all '0' or all '1' are not valid according to XGAKIT source-code, in fact bit0=0, bit1=0, bit2=0, bit3=0 there is comment in the source-code stating "illegal value, returned under Win 3.0". I have high hopes, that fixing "Display ID and Comparator Register (Index 52)" value to some good and proper value will fix XGAAIDOS.SYS issue as well, because it seems I was on the right track here:

https://github.com/86Box/86Box/issues/2968#issuecomment-1356390801

because according to IBM own document:

https://archive.org/details/bitsavers_ibmpccardseferenceManualMay92_1756350

the "display ID" is set via "“XGA Display Mode Query and Set (DMQS)”.

**about PR2 :** I expect something like PR1, i.e. some wrong bit, because it seems XGAKIT is more strict than  IBM BIOS/Reference Diskette when parsing each bit meaning. I haven't work on this problem and so I don't know for sure at the moment

**about PR3 :** currently have no idea what's wrong with that, having looked at that part of the source code at all

So, step-by-step, let first fix 86Box with proper "Display ID and Comparator Register (Index 52)" value, then maybe even XGAAIDOS.SYS will start to work and proceed with the rest of the problems from there. Thank you!

P.S. Two notes I forgot to make:

1. very good info about XGA POS_Registers I used during my investigation of PR1: https://ardent-tool.com/video/Programming_XGA_Video_POS_Registers.html
2. XGAKIT source-code is almost entirely IBM-made and from their document "Preliminary XGA Video Subsystem Hardware Users Guide" - unfortunately that IBM document seems lost, i.e. not available anymore. So, that's why XGAKIT source-code is very valuable source of information how the things are supposed to be.

**Comment by matt713** [comment]
I am confused now, I've just checked:
https://github.com/86Box/86Box/blob/master/src/video/vid_xga.c

and 

`                case 0x52:
                    ret = 0x0b;`

which should be OK, but XGABIT receives from 86Box value of 0xff for that same Index Register. So, what is going on exactly? I mean why XGABIT gets 0xff when try to read that value from 86box even it seems correct in the 86box source-code? I know that for sure, because as I mentioned in my previous post I hard-coded it in XGABIT to 0xfb, which essentially is the same 'b' value for the least significant 4 bits.

**Comment by matt713** [comment]
I am even further puzzled. So, setup msys 86box build environment, build debug version of 86box - no problem with that, modify line "**pclog("# VM: %s\n#\n", vm_name);**" to "**pclog("# MATT TEST VM: %s\n#\n", vm_name);**" in 86box.c to make sure "pclog()" works, indeed it does:

```
$ ./86Box.exe
#
# 86Box v3.11 logfile, created 2022/12/22 14:16:38
#
# **MATT TEST VM**: src

```

but, yet **not a single pclog() call** from "vid_xga.c" even I **UNcommented** all current such calls in the source code. I would really appreciate some help. How to get debug pclog messages from "vid_xga.c"?

**Comment by matt713** [comment]
> Hmm.

exactly, it's like "vid_xga.c" code is not executed at all, but yet, XGAKIT (with hard-coded "Display ID and Comparator Register (Index 52)" read, instead the 0xf it gets from 86box for some reason), manages to initialize XGA 132-characters per line text mode, i.e. that suggest there is some XGA emu running.

**Comment by matt713** [comment]
So, I get "pclog" output from 8514A emu without any problems (from 1st try):

`IBM 8514/A: VGA ON = 1, val = 00
IBM 8514/A: DISP_CNTL write 22E8 = 52, SCANMODULOS = 0
IBM 8514/A: VGA ON = 0, val = 07
1024x768 clock mode, hdisp = 1024, htotal = 1, vtotal = 408, vsyncstart = 384, interlace = 01
8514 enabled, hdisp=0, vtotal=0, htotal=0, dispend=384, rowoffset=128, split=0, vsyncstart=0, split=00000000
1024x768 clock mode, hdisp = 1024, htotal = 1, vtotal = 408, vsyncstart = 384, interlace = 01
`

but still nothing from "vid_xga.c"... It's really like "vid_xga.c" code is **not Running** in my setup, why I don't know. However, that also explains why read of "Display ID and Comparator Register (Index 52)" in XGAKIT returns 0xf, even in "vid_xga.c" source code it's clearly set it to 0x0b. I guess the fight continues to make sense what is going on...

**Comment by matt713** [comment]
Add many my own "pclog" to "vid_xga.c" - now I get one line output:

`001_xga_mca_write_io_removehandler`

which is my "pclog" in xga_mca_write() function just before io_removehandler() is called:

`pclog("001_xga_mca_write_io_removehandler\n");
io_removehandler(0x2100 + (xga->instance << 4), 0x0010, xga_ext_inb, NULL, NULL, xga_ext_outb, NULL, NULL, svga);
mem_mapping_disable(&xga->bios_rom.mapping);
mem_mapping_disable(&xga->memio_mapping);`

So, xga_init() function in my case is never called, hence xga_ext_inb() and xga_ext_outb() are never called. Anyone, any ideas?

**Comment by TC1995** [comment]
I have uncommitted changes pending, but they are not ready yet.

**Comment by matt713** [comment]
> I have uncommitted changes pending, but they are not ready yet.

if they are not to hard to apply to my building environment, please, share a patch here - I can patch my code, build and test. Now, I get good experience in building 86box and working with "pclog" to get at least some sense of what is going on.

**Comment by TC1995** [comment]
well my patch is too large at the moment.

**Comment by matt713** [comment]
ok, thanks, I will put it on hold for the time being, waiting for your patch to be merged to the main repository.

**Comment by matt713** [comment]
just out of curiosity I did some more work to pin-point the exact root cause. So, the reason is POS Register 2, value is set to 0x0c, as result of that, the following check fails:

`if (xga->pos_regs[2] & 1) {`

and that leads to the following io_sethandler() is never called:

`io_sethandler(0x2100 + (xga->instance << 4), 0x0010, xga_ext_inb, NULL, NULL, xga_ext_outb, NULL, NULL, svga);`

hence why the reads to 0x216x I mentioned here:

https://github.com/86Box/86Box/issues/2983#issuecomment-1362013899

return wrong 0x0f value instead good value (as set in vid_xga.c) of 0x0b. There are probably more problems, but that is from where it starts - I hope it's already addressed in the upcoming @TC1995 patches. I am writing this just for completeness, because at least now I get it why "Display ID and Comparator Register (Index 52)" read that XGAKIT is doing - fails with undefined 0x0f value, it's just because the io_sethandler() for 0x216x is currently never called in vid_xga.c.

**Comment by matt713** [comment]
@TC1995 

I hope you see this and tell me your opinion, because I am not quite sure where to post it. So, it's a little bit off-topic, but at the same time directly connected - it's about why "XGAAIDOS.sys" fails. I did investigate today, and better start with 3 screenshots, because a picture is a thousand words:
* here check the comment starting with "Matt"
![01_xga_ai_sys](https://user-images.githubusercontent.com/59395613/209408181-fffcc41c-d894-4222-9f73-8a789c6ba1a9.png)
* very interesting check, I am going to write small Proof-of-Concept application in ASM that calls the same Dos INT 15 (AH = C0h) "System - Get Configuration" for investigation why it fails. It took me several hours to understand what is doing, because Dos INT 15 (AH = C0h) is very poorly documented, but finally found PS2 documentation about it here: https://stanislavs.org/helppc/int_15-c0.html and based on it now makes perfect sense - it is check if the Machine has MCA bus:
![02_xga_ai_sys](https://user-images.githubusercontent.com/59395613/209408195-0350b725-e875-4461-972d-7fb938b3346a.png)
* this is call to Dos INT 15 (AH = C4h) for getting MCA Base POS Register address and other POS-related tasks: https://stanislavs.org/helppc/int_15-c4.html it also seems to me fails, again I am going to write small Proof-of-Concept application in ASM that makes the same call for further investigation.
![03_xga_ai_sys](https://user-images.githubusercontent.com/59395613/209408199-1e8dfb04-7e92-4b16-8b22-d4ccaa4a834c.png)

So, making "XGAAIDOS.sys" to work is really tremendous tasks, i.e. all those Dos INT calls need to pass. What you think? I think I am on the correct path, but it's good to hear other other options.

**Comment by TC1995** [comment]
I honestly don't know at this point about xgaaidos.sys.

**Comment by matt713** [comment]
> I honestly don't know at this point about xgaaidos.sys.

no worries, I will continue investigate it, 2nd screenshot, at least after read https://stanislavs.org/helppc/int_15-c0.html is obvious, because "[bx+5]" bit is the "Featured byte" and "test al, 2" is check bit1 and that's " 0=PC bus, 1=Micro Channel". So, IBM put insane number of checks for everything, before load it. I want waste your time, just get back, when make ASM Proof-of-Concept tools that make the same check and confirmed which exactly of them fail in 86box.

**Comment by TC1995** [comment]
Well, pos2 bit 1 should be modified only by the bios.

**Comment by OBattler** [comment]
It's executing INT 15h, AX = C401h, so AL = 01. I'm looking for the documentation on Ralf Brown's interrupt list here: http://www.ctyme.com/intr/rb-1608.htm . AL = 01 is enable selected slot for setup, with the slot in BL. BL is set to CL, which is 9 (see the mov cx,9 earlier). Then it's decreased. So it's trying to access slot 8 for setup. Now, 8 means Adapter Setup, with the actual slot number in the lower 3 bits, in this case, (8 & 7 = 0), so slot 0.

Seems it's doing this in a loop, I suspect there must be an inc cx or something somewhere in the end of this loop, and then if it's above ((8 + 7) + 1) = 16, get out of the loop. That way, it would loop through all the slots. I think it attempts to somehow identify the XGA. Which would mean that it's failing to identify the emulated XGA as such.

It also does execute INT 15h, AX = C400h before that, to get the POS registers base address. That makes sense.

**Comment by matt713** [comment]
@TC1995 
Yes, but just to double check I already made small tool in x86 assembler - called it "ps2bus.asm" to make such check (basically copy the ASM code from "XGAAIDOS.sys" and add message prints based on the check) - it outputs "PC Bus" or "MC Bus" - I will share the code (not doing it at the moment, because it's a mess, I want to clean it up first) as it can be used for template to do independent check of any suspecting ASM code from "XGAAIDOS.sys". It's also very easy to build, as I used Turbo Assembler.

@OBattler 
I completely agree with analysis - it's the same as mine. So, XGAKIT is doing almost the same XGA checks as "XGAAIDOS.sys" :

http://www.dcee.net/Files/Programm/C/xgakit.arj

and kind of confirm XGAKIT's authors words that basically all XGAKIT's code is taken from IBM "Preliminary XGA Video Subsystem Hardware Users Guide" document.

And yes, in XGAKIT code is very obvious how CX is used for Loop-counter to do thought the 8 slots.

I am doing separate tool in x86 assembler to do the same INT 15h, AX = C401h calls to see which one fails. 

So, what still makes sense is first XGA emu be fixed to work with XGAKIT - as XGAKIT is simpler and opensource (i.e. easy to read and debug), XGAKIT has strict XGA checks, but less strict and then "XGAAIDOS.sys", which is complex and XGA checks are very strict.

In any case I am very optimistic for success, because now we kind of know in what we're exactly looking...

**Comment by OBattler** [comment]
I think it's failing the monitor check - it expects XGA register 52h to return either 00h (apparently, an invalid value returned under Windows 3.0) or 0Fh, but 86Box returns 0Bh.

**Comment by TC1995** [comment]
by any chance, does this demo require himem?

**Comment by matt713** [comment]
> Edit: Nevermind, it expects the value to not be either 00h or 0Fh.

yes, 0x0 or 0xf are the two Invalid values, which are checked. So, any value different than those is good, i.e. 0x0b is not just good, but it means "Color Monitor" plus High-Resolution Monitor".

> by any chance, does this demo require himem?

don't know, but in my DOS installation "DEVICE = C:\C700\BIN\HIMEM.SYS" is installed and loaded ("C700" is Microsoft C Compiler 7.0 installation that did that by default).

My latest finding is **very very puzzling**: I did patch the MCA bus detection code in XGAAIDOS.sys, instead to rely on what the INT 15 (AH = C0h) call returns, i.e. on the value "mov al, es:[bx+5]", just hard-code it to either "2" or "0", which respectively according to:

https://stanislavs.org/helppc/int_15-c0.html 

is "MCA Bus present" and "No MCA, i.e. PC bus" and here is the weird results:

*  "mov al, es:[bx+5]" --> "mov al, 2", i.e. "MCA Bus present": there is no change in the execution according to "pclog" from xga_vid.c 
![set_al02](https://user-images.githubusercontent.com/59395613/209463442-2bae4e4f-780f-414f-8df5-5e8fcab1d405.png)

* "mov al, es:[bx+5]" --> "mov al, 0", i.e. "No MCA, i.e. PC bus": XGAAIDOS.sys code proceeds further with execution and there are even 6 more and good POS Reads done, here is the pclog:

> [04EC:0000F62B]: POS Read Port = 0, val = da
[04EC:0000F62B]: POS Read Port = 1, val = 8f
[04EC:0000F62B]: POS Read Port = 3, val = 6c
[04EC:0000F62B]: POS Read Port = 0, val = da
[04EC:0000F62B]: POS Read Port = 1, val = 8f
[04EC:0000F62B]: POS Read Port = 3, val = 6c

So, it makes no any sense to me - it does POS Reads, when basically INT 15 (AH = C0h) call is "faked" to return "no MCA bus" and goes straight to "(-XGA)"", i.e. no XGA, when  INT 15 (AH = C0h) call is "faked" to return "MCA Bus present". I don't know - could it be the documentation:

https://stanislavs.org/helppc/int_15-c0.html 

is wrong and bit1 of 0 to mean MCA bus present? Any ideas? I guess only dump from someone with real PS/2 and XGA card running those calls, can answer...

**Comment by matt713** [comment]
just to double-confirm myself, in the ASM code of XGAAIDOS.sys where it jumps to "(-XGA)", i.e. no XGA, I patched XGAAIDOS.sys with replacing the address of "(-XGA)" string with the address of string "(-80386)", which is the string printed when no 386 is detected and as the screenshots below of before and after shows indeed I correctly pin-point where the code jumps to "no XGA" and so it's not like I am looking in a wrong place, but still my last post with MCA check makes 0 sense to me:

* before the above patch

![noXGA](https://user-images.githubusercontent.com/59395613/209464804-004109b6-cf2b-42df-96f4-c88f931d2d34.png)


* after the above patch

![no386](https://user-images.githubusercontent.com/59395613/209464805-7962d719-251e-4dc8-9b54-ec7de0f14a7e.png)

**Comment by matt713** [comment]
maybe, the results in my last 2 posts make any sense only if "cs:word_10039" is used for Global variable to Indicate current XGA detect Status, i.e. "0" --> NoXGA, "1" --> XGA detected. if that's correct assumption then when place "cs:word_10039" is set depends on call to:

> mov     ax, 1F00h
> seg000:85EC                 int     10h             ; - VIDEO -

Googled that and got very very interesting hit - the ASM code here:

https://svn.fractint.net/branches/ssg/fractint/video.asm

> xga_detect      proc    uses es di si

        cmp     xga_reg_base,-2         ; has the XGA detector already failed?
        jne     xga_sk1                 ; ne = not yet
        jmp     xga_notfound            ; e = yes, fail again
xga_sk1:cmp     xga_reg_base,-1         ; have we already found the XGA?
        je      xga_loc                 ; e = no
        jmp     xga_found               ; yes, process it

xga_loc:push    bp                      ; save around int 10H calls
        mov     ax,1f00h                ; XGA-2 detect:
        int     10h                     ;  get DMQS length

**Comment by matt713** [comment]
**Finally, something consistent!** Coded small tool in ASM that calls "INT 10h, AX=1f00" and the call fails - see the screenshot below. **So, I cannot find any information for "INT 10h, AX=1f00" - anyone knows why it fails and how the 86box VM is supposed to be fixed that call to pass?**
![int10_ax1f00](https://user-images.githubusercontent.com/59395613/209467940-889a4d9f-5d63-49f8-8cd5-a170523b7f23.png)

**Comment by TC1995** [comment]
Could you send me your small tool? I'd like to test it in my changes.

**Comment by matt713** [comment]
> Could you send me your small tool? I'd like to test it in my changes.

yes, as I mentioned it few post before - I use the same messy ASM code I made as template and I am modifying in order to make the check I am interested in.  So, I I need to clean it now to the bare minimum for "INT 10h, AX=1f00" and share it or maybe just send you the EXE... that's faster, I already uploaded the EXE here:

https://www.mediafire.com/file/go3tg43rxmuofw3/PS2TEST.zip/file

My prediction is it will fail, because what fails doesn't exactly depend on "VID_XGA.c" but on some I guess BIOS calls that fail in 86box and that are otherwise valid only for PS/2, because I cannot find any documentation for what "INT 10h, AX=1f00" call is supposed to do.

So, the obvious current strategy is one by one in such small tools we test all those suspecting INT X calls. Currently, "INT 15 (AH = C0h)" passes, even I am not sure if '0' or '1' means MCA Present based on what I said 2-3 comment ago. "INT 10h, AX=1f00" - fails and that's the biggest clue at the moment, i.e. why it fails is the current question.

**Comment by matt713** [comment]
OK, I cannot find anything about  "INT 10h, AX=1f00", because it's referred as  "INT 10h, AH=1f" instead. Indeed it's only mentioned by IBM in:

https://archive.org/details/bitsavers_ibmpccardseferenceManualMay92_1756350

> | The recommended method of obtaining the display ID is by use of
> | a BIOS call, Int 10h, (AH) = 1Fh - XGA Display Mode Query and Set
> | (DMQS). See “XGA Display Mode Query and Set (DMQS)” on
> | page 3-192. If it is necessary to read the display ID explicitly then
> | the following procedure must be followed.

> Preliminary Draft May 19th 1992
> | Video BIOS Int 10h Software Interrupt function
> | (AH) = 1FH - XGA Display Mode Query and Set (DMQS)
> | (AL) = 00H - Read DMQS Data Length
> | On Return:
> | (AL) = 1FH - function supported
> | (BX) = Number of bytes of DMQS data

Our "cmp al, 1fh" fails, which means "function supported" or what I output as "Matt's DOS INT 10 (AX = 1f00h) PoC: FAIL". I start to think - **Could it be 86Box PS/2 BIOSes are Older, i.e. before XGA support was added by IBM to the BIOS?**

**Comment by matt713** [comment]
OK, I cannot find anything about  "INT 10h, AX=1f00", because it's referred as  "INT 10h, AH=1f" instead. Indeed it's only mentioned by IBM in:

https://archive.org/details/bitsavers_ibmpccardseferenceManualMay92_1756350

> | The recommended method of obtaining the display ID is by use of
> | a BIOS call, Int 10h, (AH) = 1Fh - XGA Display Mode Query and Set
> | (DMQS). See “XGA Display Mode Query and Set (DMQS)” on
> | page 3-192. If it is necessary to read the display ID explicitly then
> | the following procedure must be followed.

> Preliminary Draft May 19th 1992
> | Video BIOS Int 10h Software Interrupt function
> | (AH) = 1FH - XGA Display Mode Query and Set (DMQS)
> | (AL) = 00H - Read DMQS Data Length
> | On Return:
> | (AL) = 1FH - function supported
> | (BX) = Number of bytes of DMQS data

Our "cmp al, 1fh" fails, which means "function supported" or what I output as "Matt's DOS INT 10 (AX = 1f00h) PoC: FAIL". I start to think - **Could it be 86Box PS/2 BIOSes are Older, i.e. before XGA support was added by IBM to the BIOS?**

**Comment by TC1995** [comment]
Well, official XGA support in the PS/2 machines was first introduced with the 8585 or 9595 machines iirc.

**Comment by matt713** [comment]
> Well, official XGA support in the PS/2 machines was first introduced with the 8585 or 9595 machines iirc.

I am using IBM Reference disk for Model 70 Updated with XGA Options, took it from 86box Github here:

https://github.com/86Box/86Box/files/7573520/rf7080a.zip

and the XGA Options update is here:

https://github.com/86Box/86Box/files/7573521/xgaopt.zip

So, there is XGA for  Model 70, but who knows - maybe running the above does nothing in 86Box, i.e. it fails to do what is supposed to do?

BTW, the aforementioned IBM document describes the XGA detection algorithm step-by-step in details.

So, I clearly see now those steps both in "XGAAIDOS.sys" disassembler code and in "Fractint" application code - so in short and paraphrasing as how I understood IBM document:

1) check "INT 10h, AH=1f" - if it fails do manual configuration - it fails in our case (86box) and so we can assume it goes to manual mode configuration, i.e. go to step 2)

2) do POS register reads to locate the XGA Adapter  -those seems to succeed in our case if you read 2nd part of my comment here:

https://github.com/86Box/86Box/issues/2983#issuecomment-1364654740

and so it goes to step 3)

3) "To place the system board video subsystem in manual setup mode configuration, write hex 0DF to port 94H; to enable it, write hex 0FF to port 94H"

So, either we don't even reach step 3) or it fails somewhere before that.

I don't see any "pclog" output for "port 94H", but that's again I think not related to the XGA_VID.c, but somewhere else, because port 94H is described as "enable the motherboard for setup". So, I think something with PS/2 emulation in 86box is wrong, before it even reaches phase to talk to XGA_VID.c

**Comment by TC1995** [comment]
xgaaidos.sys loads fine here now.

**Comment by matt713** [comment]
> xgaaidos.sys loads fine here now.

good to hear, what about running XGADEMO? or XGAKIT?

**Comment by TC1995** [comment]
xgademo runs, xgakit too but the latter (xgakit) doesn't display properly (default is a blank 1024x768x8bpp screen).

**Comment by matt713** [comment]
> xgademo runs, xgakit too but the latter (xgakit) doesn't display properly (default is a blank 1024x768x8bpp screen).

that's huge progress, interesting, what change made it work, i.e. what of those checks i investigated here is passing now.

**Comment by matt713** [comment]
@TC1995 

Please, run this small tool I made:

https://www.mediafire.com/file/iewtmee5d710fe4/COMPID.zip/file

to see how IBM BIOS call IDs your system. Here is screenshot in my case:

![compid](https://user-images.githubusercontent.com/59395613/209471263-89796a64-94c6-436a-8e1e-1684a6aceb00.png)

**Comment by TC1995** [comment]
it returns model 80 in my case.

**Comment by TC1995** [comment]
![Monitor_1_20221225-152250-786786](https://user-images.githubusercontent.com/9269715/209471751-81fa4166-70f2-47a4-92ce-5efb2d858b30.png)

**Comment by matt713** [comment]
OK, thanks, I am going to setup such PS2 Model80 machine and re-run, all tools I made with different checks. Maybe, huge part of the problem is Model70 BIOSes included with 86box. In fact if that turns out to be true I can combine all those tools in one that will allow people to check they are running proper environment.

**Comment by matt713** [comment]
here are how my 3 tools look in the Model 80 PS2 machine I setup - first 2 are OK (model is 80 and Bus is MCA), but "INT 10h, AH=1f" still fails - anyway, I will wait for your patches and try to understand why it succeeds and fails for me currently
![myps2m80](https://user-images.githubusercontent.com/59395613/209472427-fd4cb088-6f69-4f73-ac2c-4a2b701e6c34.png)
.

**Comment by matt713** [comment]
just for completeness and so we have everything in one place - I failed to built "Fractint" from source, but here it's already build version with XGA support:

https://www.fractint.net/ftp/archive/1820/frain182.zip

in my environment it hangs very similarly to XGAKIT with last debug messages from XGA_VID.c I see:

> [0AA8:00001672]: EXT OUTB = 2168, val = 00
> [0AA8:00001977]: EXT OUTB = 216a, val = 64
> [0AA8:00001977]: EXT OUTB = 216b, val = ff

in any way it's another good and ready application to test the XGA emu, but I guess as most complicated it's the last after the emu is mature and XGAAIDOS and XGAKIT are running.

**Comment by matt713** [comment]
@TC1995 

Thanks to volunteer with two PS/2 with XGA-1 and XGA-2 card and I myself writing several test tools, I figured out the undocumented Index 53H, which is one of the first things XGAAIDOS.sys reads and exits if it doesn't like the value. So, if you want to read the whole saga, it's here:

https://www.vogons.org/viewtopic.php?p=1127947#p1127947

but what is the end result - there are only 3 valid values for Index 53H:

* 0x30 : XGA-1 card
* 0x50 : XGA-2 card with 45MHz Max "PEL" rate or "PEL" frequency
* 0x53 : XGA-2 card with 90MHz Max "PEL" rate or "PEL" frequency

So, IBM hide the read of Index 53H via DMQS call, but those currently as you know fail in 86box. So, maybe if "INT 10h, AH=1f" wasn't failing in 86box, XGAAIDOS.sys will not read Index 53H directly. 

Also, another volunteer with Model 80 already reported working DMQS calls - but I still haven't confirmed if that Model 80 machine has the same or different BIOS version than the one included with 86box.

**Comment by TC1995** [comment]
Heads up, I got it to recognize 1mb of vram.

**Comment by TC1995** [comment]
This should hopefully make it run at least.
